### PR TITLE
ZOMLINFRA-141: Add in remote fetching of origin/template branch

### PR DESCRIPTION
**Pull request by alexlatchford**
Needed for [this initial PR](https://github.com/zillow/battenberg-gitlab/pull/1) for `battenberg-gitlab`.

When you apply an upgrade you want to apply it on top of any existing `template` branch. If you're working in a fresh clone you may not have fetched any existing `origin/template` remote branch locally before upgrading. Now this does this for you!

Todo:

- [x] Fix tests

### CODE REVIEW

File: battenberg/utils.py [hunk 1]
```diff
@@ -22,3 +23,12 @@ def open_or_init_repository(path):
...
     return repo
+
+
+def construct_keypair(public_key_path: str = None, private_key_path: str = None) -> Keypair:
+    ssh_path = os.path.join(os.path.expanduser('~'), '.ssh')
+    if not public_key_path:
+        public_key_path = os.path.join(ssh_path, 'id_rsa.pub')
+    if not private_key_path:
+        private_key_path = os.path.join(ssh_path, 'id_rsa')
+    return Keypair("git", public_key_path, private_key_path, "")
```
Comment (shoelsch) [inline battenberg/utils.py]: Should we be exposing the passphrase as an argument as well?

File: battenberg/core.py [hunk 2]
```diff
@@ -171,9 +174,19 @@ def upgrade(self, checkout: str = 'master', extra_context: Dict = None, no_input
         if extra_context is None:
             extra_context = {}
 
-        # Assert template branch exist or raise an error
         if not self.is_installed():
-            raise TemplateNotFoundException()
+            # First try to pull it from the remote origin/TEMPLATE_BRANCH
+            keypair = construct_keypair()
+            self.repo.remotes['origin'].fetch([TEMPLATE_BRANCH],
+                                            callbacks=RemoteCallbacks(credentials=keypair))
```
Comment (shoelsch) [inline battenberg/core.py]: Nit: alignment looks off

File: HISTORY.md [hunk 3]
```diff
@@ -1,5 +1,9 @@
 # Release History
 
+## 0.2.0 (???)
```
Comment (alexlatchford) [inline HISTORY.md]: - [ ] TODO Set the release date before merging.

File: .travis.sh [hunk 4]
```diff
@@ -0,0 +1,21 @@
+#!/bin/sh
```
Comment (alexlatchford) [inline .travis.sh]: I just decided to try out adding support for Python v3.8 looks like we need to compile our own `libgit` for Travis CI to work but now we're successfully testing `battenberg` on 3.6, 3.7 & 3.8 😄

File: battenberg/core.py [hunk 5]
```diff
@@ -33,6 +35,16 @@ def __init__(self, repo: Repository):
     def is_installed(self) -> bool:
         return TEMPLATE_BRANCH in self.repo.listall_branches()
 
+    def fetch_remote_template(self):
```
Comment (tmckay) [inline battenberg/core.py]: Should public methods like this have docstrings?

Comment (alexlatchford) [reply battenberg/core.py]: Good point. Explicitly updated the methods to prefix private methods with `_` and added docstrings to all public methods using Google Style Guide format :)

**Pull request merged**